File: BotsSDK/korebot/src/main/java/kore/botssdk/views/coverflow/CoverFlowLayoutManger.java
package kore.botssdk.views.coverflow;

import android.animation.Animator;
import android.animation.ValueAnimator;
import android.graphics.ColorMatrix;
import android.graphics.ColorMatrixColorFilter;
import android.graphics.Paint;
import android.graphics.Rect;
import android.util.SparseArray;
import android.util.SparseBooleanArray;
import android.view.View;
import android.view.ViewGroup;
import android.view.animation.DecelerateInterpolator;

import androidx.recyclerview.widget.RecyclerView;

public class CoverFlowLayoutManger extends RecyclerView.LayoutManager {
    private final static int SCROLL_TO_RIGHT = 1;
    private final static int SCROLL_TO_LEFT = 2;
    private final static int MAX_RECT_COUNT = 100;
    private int mOffsetAll = 0;
    private int mDecoratedChildWidth = 0;
    private int mDecoratedChildHeight = 0;
    private float mIntervalRatio = 0.5f;
    private int mStartX = 0;
    private int mStartY = 0;
    private SparseArray<Rect> mAllItemFrames = new SparseArray<>();
    private SparseBooleanArray mHasAttachedItems = new SparseBooleanArray();
    private RecyclerView.Recycler mRecycle;
    private RecyclerView.State mState;
    private ValueAnimator mAnimation;
    private int mSelectPosition = 0;
    private int mLastSelectPosition = 0;
    private OnSelected mSelectedListener;
    private boolean mIsFlatFlow = false;
    private boolean mItemGradualGrey = false;
    private boolean mItemGradualAlpha = false;
    private boolean mIsLoop = false;
    private boolean mItem3D = false;

    private CoverFlowLayoutManger(boolean isFlat, boolean isGreyItem, boolean isAlphaItem, float cstInterval, boolean isLoop, boolean is3DItem) {
        mIsFlatFlow = isFlat;
        mItemGradualGrey = isGreyItem;
        mItemGradualAlpha = isAlphaItem;
        mIsLoop = isLoop;
        mItem3D = is3DItem;
        if (cstInterval >= 0) {
            mIntervalRatio = cstInterval;
        } else {
            if (mIsFlatFlow) {
                mIntervalRatio = 1.1f;
            }
        }
    }

    @Override
    public RecyclerView.LayoutParams generateDefaultLayoutParams() {
        return new RecyclerView.LayoutParams(ViewGroup.LayoutParams.WRAP_CONTENT, ViewGroup.LayoutParams.WRAP_CONTENT);
    }

    @Override
    public void onLayoutChildren(RecyclerView.Recycler recycler, RecyclerView.State state) {
        if (getItemCount() <= 0 || state.isPreLayout()) {
            mOffsetAll = 0;
            return;
        }
        mAllItemFrames.clear();
        mHasAttachedItems.clear();

        View scrap = recycler.getViewForPosition(0);
        addView(scrap);
        measureChildWithMargins(scrap, 0, 0);
        mDecoratedChildWidth = getDecoratedMeasuredWidth(scrap);
        mDecoratedChildHeight = getDecoratedMeasuredHeight(scrap);
        mStartX = Math.round((getHorizontalSpace() - mDecoratedChildWidth) * 1.0f / 2);
        mStartY = Math.round((getVerticalSpace() - mDecoratedChildHeight) * 1.0f / 2);

        float offset = mStartX;

        for (int i = 0; i < getItemCount() && i < MAX_RECT_COUNT; i++) {
            Rect frame = mAllItemFrames.get(i);
            if (frame == null) {
                frame = new Rect();
            }
            frame.set(Math.round(offset), mStartY, Math.round(offset + mDecoratedChildWidth), mStartY + mDecoratedChildHeight);
            mAllItemFrames.put(i, frame);
            mHasAttachedItems.put(i, false);
            offset = offset + getIntervalDistance();
        }

        detachAndScrapAttachedViews(recycler);
        if ((mRecycle == null || mState == null) && mSelectPosition != 0) {
            mOffsetAll = calculateOffsetForPosition(mSelectPosition);
            onSelectedCallBack();
        }

        layoutItems(recycler, state, SCROLL_TO_LEFT);

        mRecycle = recycler;
        mState = state;
    }

    @Override
    public int scrollHorizontallyBy(int dx, RecyclerView.Recycler recycler, RecyclerView.State state) {
        if (mAnimation != null && mAnimation.isRunning()) mAnimation.cancel();
        int travel = dx;
        if (!mIsLoop) {
            if (dx + mOffsetAll < 0) {
                travel = -mOffsetAll;
            } else if (dx + mOffsetAll > getMaxOffset()) {
                travel = (int) (getMaxOffset() - mOffsetAll);
            }
        }
        mOffsetAll += travel;
        layoutItems(recycler, state, dx > 0 ? SCROLL_TO_LEFT : SCROLL_TO_RIGHT);
        return travel;
    }

    private void layoutItems(RecyclerView.Recycler recycler, RecyclerView.State state, int scrollDirection) {
        if (state == null || state.isPreLayout()) return;

        Rect displayFrame = new Rect(mOffsetAll, 0, mOffsetAll + getHorizontalSpace(), getVerticalSpace());

        int position = 0;
        for (int i = 0; i < getChildCount(); i++) {
            View child = getChildAt(i);
            if (child.getTag() != null) {
                TAG tag = checkTag(child.getTag());
                position = tag.pos;
            } else {
                position = getPosition(child);
            }

            Rect rect = getFrame(position);

            if (!Rect.intersects(displayFrame, rect)) {
                removeAndRecycleView(child, recycler);
                mHasAttachedItems.delete(position);
            } else {
                layoutItem(child, rect);
                mHasAttachedItems.put(position, true);
            }
        }

        if (position == 0) position = getCenterPosition();

        int min = position - 20;
        int max = position + 20;

        if (!mIsLoop) {
            if (min < 0) min = 0;
            if (max > getItemCount()) max = getItemCount();
        }

        for (int i = min; i < max; i++) {
            Rect rect = getFrame(i);
            if (Rect.intersects(displayFrame, rect) &&
                    !mHasAttachedItems.get(i)) {
                int actualPos = i % getItemCount();
                if (actualPos < 0) actualPos = getItemCount() + actualPos;

                View scrap = recycler.getViewForPosition(actualPos);
                checkTag(scrap.getTag());
                scrap.setTag(new TAG(i));

                measureChildWithMargins(scrap, 0, 0);
                if (scrollDirection == SCROLL_TO_RIGHT || mIsFlatFlow) {
                    addView(scrap, 0);
                } else {
                    addView(scrap);
                }
                layoutItem(scrap, rect);
                mHasAttachedItems.put(i, true);
            }
        }
    }

    private void layoutItem(View child, Rect frame) {
        layoutDecorated(child, frame.left - mOffsetAll, frame.top, frame.right - mOffsetAll, frame.bottom);
        if (!mIsFlatFlow) {
            child.setScaleX(computeScale(frame.left - mOffsetAll));
            child.setScaleY(computeScale(frame.left - mOffsetAll));
        }

        if (mItemGradualAlpha) {
            child.setAlpha(computeAlpha(frame.left - mOffsetAll));
        }

        if (mItemGradualGrey) {
            greyItem(child, frame);
        }

        if (mItem3D) {
            item3D(child, frame);
        }
    }

    private Rect getFrame(int index) {
        Rect frame = mAllItemFrames.get(index);
        if (frame == null) {
            frame = new Rect();
            float offset = mStartX + getIntervalDistance() * index;
            frame.set(Math.round(offset), mStartY, Math.round(offset + mDecoratedChildWidth), mStartY + mDecoratedChildHeight);
        }

        return frame;
    }

    private void greyItem(View child, Rect frame) {
        float value = computeGreyScale(frame.left - mOffsetAll);
        ColorMatrix cm = new ColorMatrix(new float[]{
                value, 0, 0, 0, 120 * (1 - value),
                0, value, 0, 0, 120 * (1 - value),
                0, 0, value, 0, 120 * (1 - value),
                0, 0, 0, 1, 250 * (1 - value),
        });
        Paint greyPaint = new Paint();
        greyPaint.setColorFilter(new ColorMatrixColorFilter(cm));

        child.setLayerType(View.LAYER_TYPE_HARDWARE, greyPaint);
        if (value >= 1) {
            child.setLayerType(View.LAYER_TYPE_NONE, null);
        }
    }

    private void item3D(View child, Rect frame) {
        float center = (frame.left + frame.right - 2 * mOffsetAll) / 2f;
        float value = (center - (mStartX + mDecoratedChildWidth / 2f)) * 1f / (getItemCount() * getIntervalDistance());
        value = (float) Math.sqrt(Math.abs(value));
        float symbol = center > (mStartX + mDecoratedChildWidth / 2f) ? -1 : 1;
        child.setRotationY(symbol * 50 * value);
    }

    @Override
    public void onScrollStateChanged(int state) {
        super.onScrollStateChanged(state);
        switch (state) {
            case RecyclerView.SCROLL_STATE_IDLE:
                fixOffsetWhenFinishScroll();
                break;
            case RecyclerView.SCROLL_STATE_DRAGGING:
                break;
            case RecyclerView.SCROLL_STATE_SETTLING:
                break;
        }
    }

    @Override
    public void scrollToPosition(int position) {
        if (position < 0 || position > getItemCount() - 1) return;
        mOffsetAll = calculateOffsetForPosition(position);
        if (mRecycle == null || mState == null) {
            mSelectPosition = position;
        } else {
            layoutItems(mRecycle, mState, position > mSelectPosition ? SCROLL_TO_LEFT : SCROLL_TO_RIGHT);
            onSelectedCallBack();
        }
    }

    @Override
    public void smoothScrollToPosition(RecyclerView recyclerView, RecyclerView.State state, int position) {
        if (mIsLoop) return;
        int finalOffset = calculateOffsetForPosition(position);
        if (mRecycle == null || mState == null) {
            mSelectPosition = position;
        } else {
            startScroll(mOffsetAll, finalOffset);
        }
    }

    @Override
    public boolean canScrollHorizontally() {
        return true;
    }

    @Override
    public void onAdapterChanged(RecyclerView.Adapter oldAdapter, RecyclerView.Adapter newAdapter) {
        removeAllViews();
        mRecycle = null;
        mState = null;
        mOffsetAll = 0;
        mSelectPosition = 0;
        mLastSelectPosition = 0;
        mHasAttachedItems.clear();
        mAllItemFrames.clear();
    }

    private int getHorizontalSpace() {
        return getWidth() - getPaddingRight() - getPaddingLeft();
    }

    private int getVerticalSpace() {
        return getHeight() - getPaddingBottom() - getPaddingTop();
    }

    private float getMaxOffset() {
        return (getItemCount() - 1) * getIntervalDistance();
    }

    private float computeScale(int x) {
        float scale = 1 - Math.abs(x - mStartX) * 1.0f / Math.abs(mStartX + mDecoratedChildWidth / mIntervalRatio);
        if (scale < 0) scale = 0;
        if (scale > 1) scale = 1;
        return scale;
    }

    private float computeGreyScale(int x) {
        float itemMidPos = x + mDecoratedChildWidth / 2;
        float itemDx2Mid = Math.abs(itemMidPos - getHorizontalSpace() / 2f);
        float value = 1 - itemDx2Mid * 1.0f / (getHorizontalSpace() / 2);
        if (value < 0.1) value = 0.1f;
        if (value > 1) value = 1;
        value = (float) Math.pow(value, .8);
        return value;
    }

    private float computeAlpha(int x) {
        float alpha = 1 - Math.abs(x - mStartX) * 1.0f / Math.abs(mStartX + mDecoratedChildWidth / mIntervalRatio);
        if (alpha < 0.3f) alpha = 0.3f;
        if (alpha > 1) alpha = 1.0f;
        return alpha;
    }

    private int calculateOffsetForPosition(int position) {
        return Math.round(getIntervalDistance() * position);
    }

    private void fixOffsetWhenFinishScroll() {
        if (getIntervalDistance() != 0) {
            int scrollN = (int) (mOffsetAll * 1.0f / getIntervalDistance());
            float moreDx = (mOffsetAll % getIntervalDistance());
            if (Math.abs(moreDx) > (getIntervalDistance() * 0.5)) {
                if (moreDx > 0) scrollN++;
                else scrollN--;
            }
            int finalOffset = scrollN * getIntervalDistance();
            startScroll(mOffsetAll, finalOffset);
            mSelectPosition = Math.abs(Math.round(finalOffset * 1.0f / getIntervalDistance())) % getItemCount();
        }
    }

    private void startScroll(int from, int to) {
        if (mAnimation != null && mAnimation.isRunning()) {
            mAnimation.cancel();
        }
        final int direction = from < to ? SCROLL_TO_LEFT : SCROLL_TO_RIGHT;
        mAnimation = ValueAnimator.ofFloat(from, to);
        mAnimation.setDuration(500);
        mAnimation.setInterpolator(new DecelerateInterpolator());
        mAnimation.addUpdateListener(animation -> {
            mOffsetAll = Math.round((float) animation.getAnimatedValue());
            layoutItems(mRecycle, mState, direction);
        });
        mAnimation.addListener(new Animator.AnimatorListener() {
            @Override
            public void onAnimationStart(Animator animation) {
            }

            @Override
            public void onAnimationEnd(Animator animation) {
                onSelectedCallBack();
            }

            @Override
            public void onAnimationCancel(Animator animation) {
            }

            @Override
            public void onAnimationRepeat(Animator animation) {
            }
        });
        mAnimation.start();
    }

    private int getIntervalDistance() {
        return Math.round(mDecoratedChildWidth * mIntervalRatio);
    }

    private void onSelectedCallBack() {
        mSelectPosition = Math.round(mOffsetAll / getIntervalDistance());
        mSelectPosition = Math.abs(mSelectPosition % getItemCount());
        if (mSelectedListener != null && mSelectPosition != mLastSelectPosition) {
            mSelectedListener.onItemSelected(mSelectPosition);
        }
        mLastSelectPosition = mSelectPosition;
    }

    private TAG checkTag(Object tag) {
        if (tag != null) {
            if (tag instanceof TAG) {
                return ((TAG) tag);
            } else {
                throw new IllegalArgumentException("You should not use View#setTag(Object tag), use View#setTag(int key, Object tag) instead!");
            }
        } else {
            return null;
        }
    }

    public int getFirstVisiblePosition() {
        Rect displayFrame = new Rect(mOffsetAll, 0, mOffsetAll + getHorizontalSpace(), getVerticalSpace());
        int cur = getCenterPosition();
        for (int i = cur - 1; ; i--) {
            Rect rect = getFrame(i);
            if (rect.left <= displayFrame.left) {
                return Math.abs(i) % getItemCount();
            }
        }
    }

    public int getLastVisiblePosition() {
        Rect displayFrame = new Rect(mOffsetAll, 0, mOffsetAll + getHorizontalSpace(), getVerticalSpace());
        int cur = getCenterPosition();
        for (int i = cur + 1; ; i++) {
            Rect rect = getFrame(i);
            if (rect.right >= displayFrame.right) {
                return Math.abs(i) % getItemCount();
            }
        }
    }

    int getChildActualPos(int index) {
        View child = getChildAt(index);
        if (child.getTag() != null) {
            TAG tag = checkTag(child.getTag());
            return tag.pos;
        } else {
            return getPosition(child);
        }
    }

    public int getMaxVisibleCount() {
        int oneSide = (getHorizontalSpace() - mStartX) / (getIntervalDistance());
        return oneSide * 2 + 1;
    }

    int getCenterPosition() {
        int pos = mOffsetAll / getIntervalDistance();
        int more = mOffsetAll % getIntervalDistance();
        if (Math.abs(more) >= getIntervalDistance() * 0.5f) {
            if (more >= 0) pos++;
            else pos--;
        }
        return pos;
    }

    public void setOnSelectedListener(OnSelected l) {
        mSelectedListener = l;
    }

    public int getSelectedPos() {
        return mSelectPosition;
    }

    public interface OnSelected {
        void onItemSelected(int position);
    }

    private class TAG {
        int pos;

        TAG(int pos) {
            this.pos = pos;
        }
    }

    static class Builder {
        boolean isFlat = false;
        boolean isGreyItem = false;
        boolean isAlphaItem = false;
        float cstIntervalRatio = -1f;
        boolean isLoop = false;
        boolean is3DItem = false;

        Builder setFlat(boolean flat) {
            isFlat = flat;
            return this;
        }

        Builder setGreyItem(boolean greyItem) {
            isGreyItem = greyItem;
            return this;
        }

        Builder setAlphaItem(boolean alphaItem) {
            isAlphaItem = alphaItem;
            return this;
        }

        Builder setIntervalRatio(float ratio) {
            cstIntervalRatio = ratio;
            return this;
        }

        Builder loop() {
            isLoop = true;
            return this;
        }

        Builder set3DItem(boolean d3) {
            is3DItem = true;
            return this;
        }

        public CoverFlowLayoutManger build() {
            return new CoverFlowLayoutManger(isFlat, isGreyItem,
                    isAlphaItem, cstIntervalRatio, isLoop, is3DItem);
        }
    }
}

File: BotSdk/ui/src/main/java/com/kore/ui/views/CoverFlowLayoutManger.kt
package com.kore.ui.views

import android.animation.Animator
import android.animation.ValueAnimator
import android.annotation.SuppressLint
import android.graphics.ColorMatrix
import android.graphics.ColorMatrixColorFilter
import android.graphics.Paint
import android.graphics.Rect
import android.util.SparseArray
import android.util.SparseBooleanArray
import android.view.View
import android.view.ViewGroup
import android.view.animation.DecelerateInterpolator
import androidx.recyclerview.widget.RecyclerView
import androidx.recyclerview.widget.RecyclerView.Recycler
import kotlin.math.abs
import kotlin.math.roundToInt
import kotlin.math.sqrt

@SuppressLint("UnknownNullness")
class CoverFlowLayoutManger internal constructor(
    private val isFlatFlow: Boolean,
    private val itemGradualGrey: Boolean,
    private val itemGradualAlpha: Boolean,
    cstInterval: Float,
    private val isLoop: Boolean,
    private val item3D: Boolean
) : RecyclerView.LayoutManager() {
    var offsetAll = 0
    private var decoratedChildWidth = 0
    private var decoratedChildHeight = 0
    private var intervalRatio = 0.5f
    private var startX = 0
    private var startY = 0
    private val allItemFrames = SparseArray<Rect>()
    private val hasAttachedItems = SparseBooleanArray()
    private var recycler: Recycler? = null
    private var state: RecyclerView.State? = null
    private var animation: ValueAnimator? = null
    var selectedPos = 0
        private set
    private var lastSelectPosition = 0
    private var selectedListener: OnSelected? = null

    init {
        if (cstInterval >= 0) {
            intervalRatio = cstInterval
        } else {
            if (isFlatFlow) {
                intervalRatio = 1.1f
            }
        }
    }

    override fun generateDefaultLayoutParams(): RecyclerView.LayoutParams {
        return RecyclerView.LayoutParams(ViewGroup.LayoutParams.WRAP_CONTENT, ViewGroup.LayoutParams.WRAP_CONTENT)
    }

    override fun onLayoutChildren(recycler: Recycler, state: RecyclerView.State) {
        if (itemCount <= 0 || state.isPreLayout) {
            offsetAll = 0
            return
        }
        allItemFrames.clear()
        hasAttachedItems.clear()
        val scrap = recycler.getViewForPosition(0)
        addView(scrap)
        measureChildWithMargins(scrap, 0, 0)
        decoratedChildWidth = getDecoratedMeasuredWidth(scrap)
        decoratedChildHeight = getDecoratedMeasuredHeight(scrap)
        startX = Math.round((horizontalSpace - decoratedChildWidth) * 1.0f / 2)
        startY = Math.round((verticalSpace - decoratedChildHeight) * 1.0f / 2)
        var offset = startX.toFloat()
        var i = 0
        while (i < itemCount && i < MAX_RECT_COUNT) {
            var frame = allItemFrames[i]
            if (frame == null) {
                frame = Rect()
            }
            frame[offset.roundToInt(), startY, (offset + decoratedChildWidth).roundToInt()] = startY + decoratedChildHeight
            allItemFrames.put(i, frame)
            hasAttachedItems.put(i, false)
            offset += intervalDistance
            i++
        }
        detachAndScrapAttachedViews(recycler)
        if ((this.recycler == null || this.state == null) && selectedPos != 0) {
            offsetAll = calculateOffsetForPosition(selectedPos)
            onSelectedCallBack()
        }
        layoutItems(recycler, state, SCROLL_TO_LEFT)
        this.recycler = recycler
        this.state = state
    }

    override fun scrollHorizontallyBy(dx: Int, recycler: Recycler, state: RecyclerView.State): Int {
        if (animation != null && animation!!.isRunning) animation?.cancel()
        var travel = dx
        if (!isLoop) {
            if (dx + offsetAll < 0) {
                travel = -offsetAll
            } else if (dx + offsetAll > maxOffset) {
                travel = (maxOffset - offsetAll).toInt()
            }
        }
        offsetAll += travel
        layoutItems(recycler, state, if (dx > 0) SCROLL_TO_LEFT else SCROLL_TO_RIGHT)
        return travel
    }

    fun layoutItems(recycler: Recycler?, state: RecyclerView.State?, scrollDirection: Int) {
        if (state == null || state.isPreLayout) return
        val displayFrame = Rect(offsetAll, 0, offsetAll + horizontalSpace, verticalSpace)
        var position = 0
        for (i in 0 until childCount) {
            val child = getChildAt(i)
            if (child != null) {
                position = if (child.tag != null) {
                    val tag = checkTag(child.tag)
                    tag?.pos ?: 0
                } else {
                    getPosition(child)
                }
                val rect = getFrame(position)
                if (!Rect.intersects(displayFrame, rect)) {
                    removeAndRecycleView(child, recycler!!)
                    hasAttachedItems.delete(position)
                } else {
                    layoutItem(child, rect)
                    hasAttachedItems.put(position, true)
                }
            }
        }
        if (position == 0) position = centerPosition
        var min = position - 20
        var max = position + 20
        if (!isLoop) {
            if (min < 0) min = 0
            if (max > itemCount) max = itemCount
        }
        for (i in min until max) {
            val rect = getFrame(i)
            if (Rect.intersects(displayFrame, rect) && !hasAttachedItems[i]) {
                var actualPos = i % itemCount
                if (actualPos < 0) actualPos += itemCount
                val scrap = recycler!!.getViewForPosition(actualPos)
                checkTag(scrap.tag)
                scrap.tag = TAG(i)
                measureChildWithMargins(scrap, 0, 0)
                if (scrollDirection == SCROLL_TO_RIGHT || isFlatFlow) {
                    addView(scrap, 0)
                } else {
                    addView(scrap)
                }
                layoutItem(scrap, rect)
                hasAttachedItems.put(i, true)
            }
        }
    }

    private fun layoutItem(child: View, frame: Rect) {
        layoutDecorated(child, frame.left - offsetAll, frame.top, frame.right - offsetAll, frame.bottom)
        if (!isFlatFlow) {
            child.scaleX = computeScale(frame.left - offsetAll)
            child.scaleY = computeScale(frame.left - offsetAll)
        }
        if (itemGradualAlpha) {
            child.alpha = computeAlpha(frame.left - offsetAll)
        }
        if (itemGradualGrey) {
            greyItem(child, frame)
        }
        if (item3D) {
            item3D(child, frame)
        }
    }

    /**
     * 动态获取Item的位置信息
     *
     * @param index item位置
     * @return item的Rect信息
     */
    private fun getFrame(index: Int): Rect {
        var frame = allItemFrames[index]
        if (frame == null) {
            frame = Rect()
            val offset = (startX + intervalDistance * index).toFloat() //原始位置累加（即累计间隔距离）
            frame[offset.roundToInt(), startY, (offset + decoratedChildWidth).roundToInt()] = startY + decoratedChildHeight
        }
        return frame
    }

    private fun greyItem(child: View, frame: Rect) {
        val value = computeGreyScale(frame.left - offsetAll)
        val cm = ColorMatrix(
            floatArrayOf(
                value,
                0f,
                0f,
                0f,
                120 * (1 - value),
                0f,
                value,
                0f,
                0f,
                120 * (1 - value),
                0f,
                0f,
                value,
                0f,
                120 * (1 - value),
                0f,
                0f,
                0f,
                1f,
                250 * (1 - value)
            )
        )
        val greyPaint = Paint()
        greyPaint.colorFilter = ColorMatrixColorFilter(cm)

        // Create a hardware layer with the grey paint
        child.setLayerType(View.LAYER_TYPE_HARDWARE, greyPaint)
        if (value >= 1) {
            // Remove the hardware layer
            child.setLayerType(View.LAYER_TYPE_NONE, null)
        }
    }

    private fun item3D(child: View, frame: Rect) {
        val center = (frame.left + frame.right - 2 * offsetAll) / 2f
        var value = (center - (startX + decoratedChildWidth / 2f)) / (itemCount * intervalDistance)
        value = sqrt(abs(value).toDouble()).toFloat()
        val symbol = (if (center > startX + decoratedChildWidth / 2f) -1 else 1).toFloat()
        child.rotationY = symbol * 50 * value
    }

    override fun onScrollStateChanged(state: Int) {
        super.onScrollStateChanged(state)
        when (state) {
            RecyclerView.SCROLL_STATE_IDLE -> fixOffsetWhenFinishScroll()
            RecyclerView.SCROLL_STATE_DRAGGING, RecyclerView.SCROLL_STATE_SETTLING -> {}
        }
    }

    override fun scrollToPosition(position: Int) {
        if (position < 0 || position > itemCount - 1) return
        offsetAll = calculateOffsetForPosition(position)
        if (recycler == null || state == null) {
            selectedPos = position
        } else {
            layoutItems(recycler, state, if (position > selectedPos) SCROLL_TO_LEFT else SCROLL_TO_RIGHT)
            onSelectedCallBack()
        }
    }

    override fun smoothScrollToPosition(recyclerView: RecyclerView, state: RecyclerView.State, position: Int) {
        if (isLoop) return
        val finalOffset = calculateOffsetForPosition(position)
        if (recycler == null || this.state == null) {
            selectedPos = position
        } else {
            startScroll(offsetAll, finalOffset)
        }
    }

    override fun canScrollHorizontally(): Boolean {
        return true
    }

    override fun onAdapterChanged(oldAdapter: RecyclerView.Adapter<*>?, newAdapter: RecyclerView.Adapter<*>?) {
        removeAllViews()
        recycler = null
        state = null
        offsetAll = 0
        selectedPos = 0
        lastSelectPosition = 0
        hasAttachedItems.clear()
        allItemFrames.clear()
    }

    private val horizontalSpace: Int
        get() = width - paddingRight - paddingLeft
    private val verticalSpace: Int
        get() = height - paddingBottom - paddingTop
    private val maxOffset: Float
        get() = ((itemCount - 1) * intervalDistance).toFloat()

    private fun computeScale(x: Int): Float {
        var scale = 1 - Math.abs(x - startX) * 1.0f / Math.abs(startX + decoratedChildWidth / intervalRatio)
        if (scale < 0) scale = 0f
        if (scale > 1) scale = 1f
        return scale
    }

    private fun computeGreyScale(x: Int): Float {
        val itemMidPos = x + decoratedChildWidth.toFloat() / 2 //item中点x坐标
        val itemDx2Mid = Math.abs(itemMidPos - horizontalSpace / 2f) //item中点距离控件中点距离
        var value = 1 - itemDx2Mid / (horizontalSpace.toFloat() / 2)
        if (value < 0.1) value = 0.1f
        if (value > 1) value = 1f
        value = Math.pow(value.toDouble(), .8).toFloat()
        return value
    }

    /**
     * 计算Item半透值
     *
     * @param x Item的偏移量
     * @return 缩放系数
     */
    private fun computeAlpha(x: Int): Float {
        var alpha = 1 - abs(x - startX) * 1.0f / abs(startX + decoratedChildWidth / intervalRatio)
        if (alpha < 0.3f) alpha = 0.3f
        if (alpha > 1) alpha = 1.0f
        return alpha
    }

    private fun calculateOffsetForPosition(position: Int): Int {
        return (intervalDistance * position).toFloat().roundToInt()
    }

    private fun fixOffsetWhenFinishScroll() {
        if (intervalDistance != 0) {
            var scrollN = (offsetAll * 1.0f / intervalDistance).toInt()
            val moreDx = (offsetAll % intervalDistance).toFloat()
            if (abs(moreDx) > intervalDistance * 0.5) {
                if (moreDx > 0) scrollN++ else scrollN--
            }
            val finalOffset = scrollN * intervalDistance
            startScroll(offsetAll, finalOffset)
            selectedPos = Math.abs((finalOffset * 1.0f / intervalDistance).roundToInt()) % itemCount
        }
    }

    private fun startScroll(from: Int, to: Int) {
        if (animation != null && animation!!.isRunning) {
            animation?.cancel()
        }
        val direction = if (from < to) SCROLL_TO_LEFT else SCROLL_TO_RIGHT
        animation = ValueAnimator.ofFloat(from.toFloat(), to.toFloat())
        animation?.duration = 500
        animation?.interpolator = DecelerateInterpolator()
        animation?.addUpdateListener { animation: ValueAnimator ->
            offsetAll = (animation.animatedValue as Float).roundToInt()
            layoutItems(recycler, state, direction)
        }
        animation?.addListener(object : Animator.AnimatorListener {
            override fun onAnimationStart(animation: Animator) {}
            override fun onAnimationEnd(animation: Animator) {
                onSelectedCallBack()
            }

            override fun onAnimationCancel(animation: Animator) {}
            override fun onAnimationRepeat(animation: Animator) {}
        })
        animation?.start()
    }

    /**
     * 获取Item间隔
     */
    private val intervalDistance: Int
        get() = (decoratedChildWidth * intervalRatio).roundToInt()

    /**
     * 计算当前选中位置，并回调
     */
    fun onSelectedCallBack() {
        selectedPos = (offsetAll.toFloat() / intervalDistance).roundToInt()
        selectedPos = abs(selectedPos % itemCount)
        if (selectedPos != lastSelectPosition) selectedListener?.onItemSelected(selectedPos)
        lastSelectPosition = selectedPos
    }

    private fun checkTag(tag: Any?): TAG? {
        return if (tag != null) {
            if (tag is TAG) {
                tag
            } else {
                throw IllegalArgumentException("You should not use View#setTag(Object tag), use View#setTag(int key, Object tag) instead!")
            }
        } else {
            null
        }
    }

    val firstVisiblePosition: Int
        get() {
            val displayFrame = Rect(offsetAll, 0, offsetAll + horizontalSpace, verticalSpace)
            val cur = centerPosition
            var i = cur - 1
            while (true) {
                val rect = getFrame(i)
                if (rect.left <= displayFrame.left) {
                    return Math.abs(i) % itemCount
                }
                i--
            }
        }
    val lastVisiblePosition: Int
        get() {
            val displayFrame = Rect(offsetAll, 0, offsetAll + horizontalSpace, verticalSpace)
            val cur = centerPosition
            var i = cur + 1
            while (true) {
                val rect = getFrame(i)
                if (rect.right >= displayFrame.right) {
                    return abs(i) % itemCount
                }
                i++
            }
        }

    fun getChildActualPos(index: Int): Int {
        val child = getChildAt(index)
        return if (child != null) {
            if (child.tag != null) {
                val tag = checkTag(child.tag)
                tag!!.pos
            } else {
                getPosition(child)
            }
        } else index
    }

    val maxVisibleCount: Int
        get() {
            val oneSide = (horizontalSpace - startX) / intervalDistance
            return oneSide * 2 + 1
        }
    val centerPosition: Int
        get() {
            var pos = offsetAll / intervalDistance
            val more = offsetAll % intervalDistance
            if (abs(more) >= intervalDistance * 0.5f) {
                if (more >= 0) pos++ else pos--
            }
            return pos
        }

    fun setOnSelectedListener(l: OnSelected?) {
        selectedListener = l
    }

    interface OnSelected {
        fun onItemSelected(position: Int)
    }

    private class TAG internal constructor(var pos: Int)
    internal class Builder {
        var isFlat = false
        var isGreyItem = false
        var isAlphaItem = false
        var cstIntervalRatio = -1f
        var isLoop = false
        var is3DItem = false
        fun setFlat(flat: Boolean): Builder {
            isFlat = flat
            return this
        }

        fun setGreyItem(greyItem: Boolean): Builder {
            isGreyItem = greyItem
            return this
        }

        fun setAlphaItem(alphaItem: Boolean): Builder {
            isAlphaItem = alphaItem
            return this
        }

        fun setIntervalRatio(ratio: Float): Builder {
            cstIntervalRatio = ratio
            return this
        }

        fun loop(): Builder {
            isLoop = true
            return this
        }

        fun set3DItem(d3: Boolean): Builder {
            is3DItem = d3
            return this
        }

        fun build(): CoverFlowLayoutManger {
            return CoverFlowLayoutManger(isFlat, isGreyItem, isAlphaItem, cstIntervalRatio, isLoop, is3DItem)
        }
    }

    companion object {
        private const val SCROLL_TO_RIGHT = 1
        private const val SCROLL_TO_LEFT = 2
        private const val MAX_RECT_COUNT = 100
    }
}